torch_datasets.py

```diff
@@ -0,0 +1,194 @@
+from typing import Dict, List, cast, Tuple
+import json
+import logging
+from overrides import overrides
+
+from torch.utils.data import Dataset as TorchDataset
+from torch.utils.data import DataLoader
+from torch.utils.data import Sampler, BatchSampler, SequentialSampler, SubsetRandomSampler
+
+from allennlp.common.registrable import Registrable
+from allennlp.common.util import add_noise_to_dict_values, lazy_groups_of
+from allennlp.data.batch import Batch as AllennlpBatch
+from allennlp.data.instance import Instance
+from allennlp.data.vocabulary import Vocabulary
+from allennlp.data import Token
+from allennlp.common.file_utils import cached_path
+from allennlp.data.fields import Field, TextField, LabelField, MetadataField
+from allennlp.data.token_indexers import SingleIdTokenIndexer, TokenIndexer
+
+logger = logging.getLogger(__name__)
+
+class Dataset(TorchDataset, Registrable):
+    def __init__(self):
+        self.vocab: Vocabulary = None
+
+    def text_to_instance(self, *inputs) -> Instance:
+
+        raise NotImplementedError
+
+    def __getitem__(self) -> Instance:
+
+        raise NotImplementedError
+
+    def index_with(self, vocab: Vocabulary):
+        self.vocab = vocab
+
+
+"""
+Here we have two SNLI readers in both of the different styles.
+They are only slightly different.
+"""
+
+
+class SnliDataset(Dataset):
+    def __init__(
+        self, file_path: str, token_indexers: Dict[str, TokenIndexer] = None, lazy: bool = False
+    ) -> None:
+
+        super().__init__()
+
+        self._tokenizer = lambda x: [Token(t) for t in x.split(" ")]
+        self._token_indexers = token_indexers or {"tokens": SingleIdTokenIndexer()}
+
+        file_path = cached_path(file_path)
+        self.examples = []
+
+        for line in open(file_path, "r"):
+            example = json.loads(line)
+            if example["gold_label"] == "-":
+                # These were cases where the annotators disagreed; we'll just skip them.  It's
+                # like 800 out of 500k examples in the training data.
+                continue
+            self.examples.append(example)
```

Comment:
Probably better to have `self.examples` actually be `self.instances`, with actual `Instance` objects, because then we can cache them very easily.